CrossCompatibility/CrossCompatibility/Utility/CompatibilityProfileLoader.cs

```diff
@@ -0,0 +1,227 @@
+
+        private static readonly Lazy<CompatibilityProfileLoader> s_sharedInstance = new Lazy<CompatibilityProfileLoader>(() => new CompatibilityProfileLoader());
+
+        private readonly JsonProfileSerializer _jsonSerializer;
+
+        private readonly ConcurrentDictionary<string, Lazy<CompatibilityProfileCacheEntry>> _profileCache;
+
+        /// <summary>
+        /// A lazy-initialized static instance to allow for a shared profile cache.
+        /// </summary>
+        public static CompatibilityProfileLoader StaticInstance => s_sharedInstance.Value;
+
+        /// <summary>
+        /// Create a new compatibility profile loader with an empty cache.
+        /// </summary>
+        public CompatibilityProfileLoader()
+        {
+            _jsonSerializer = JsonProfileSerializer.Create();
+
+            // Cache keys are filenames, which must be case-insensitive in Windows and macOS
+#if NETSTANDARD2_0
+            if (System.Runtime.InteropServices.RuntimeInformation.IsOSPlatform(OSPlatform.Linux))
+            {
+                _profileCache = new ConcurrentDictionary<string, Lazy<CompatibilityProfileCacheEntry>>();
+            }
+            else
+            {
+                _profileCache = new ConcurrentDictionary<string, Lazy<CompatibilityProfileCacheEntry>>(StringComparer.OrdinalIgnoreCase);
+            }
+#else
+            _profileCache = new ConcurrentDictionary<string, Lazy<CompatibilityProfileCacheEntry>>(StringComparer.OrdinalIgnoreCase);
+#endif
+        }
+
+        /// <summary>
+        /// For a set of profile paths, retrieve those profiles,
+        /// along with the union profile for comparison.
+        /// </summary>
+        /// <param name="profileDirPath">The absolute path to the profile directory.</param>
+        /// <param name="profilePaths">Absolute paths to all profiles to load.</param>
+        /// <param name="unionProfile">The loaded union profile to compare against.</param>
+        /// <returns>A list of hydrated profiles from all the profile paths given, not necessarily in order.</returns>
+        public IEnumerable<CompatibilityProfileData> GetProfilesWithUnion(
+            DirectoryInfo profileDirPath,
+            IEnumerable<string> profilePaths,
+            out CompatibilityProfileData unionProfile)
+        {
+            IEnumerable<CompatibilityProfileCacheEntry> profileEntries = GetProfilesFromPaths(profilePaths);
+
+            unionProfile = GetUnionProfile(profileDirPath);
+
+            return profileEntries.Select(p => p.Profile);
+        }
+
+        /// <summary>
+        /// Clear all loaded profiles from this loader.
+        /// </summary>
+        public void ClearCache()
+        {
+            _profileCache.Clear();
+        }
+
+        private IEnumerable<CompatibilityProfileCacheEntry> GetProfilesFromPaths(IEnumerable<string> profilePaths)
```

Comment:
I spent most of yesterday working on this logic in an attempt to make it safe to call in parallel and also efficient. I'm fairly confident that it does the job now, but it needs some scrutiny.